lib/sass/script/functions.rb

```diff
@@ -2078,17 +2092,51 @@ def map_get(map, key)
     # @example
     #   map-merge(("foo": 1), ("bar": 2)) => ("foo": 1, "bar": 2)
     #   map-merge(("foo": 1, "bar": 2), ("bar": 3)) => ("foo": 1, "bar": 3)
+    #   map-merge(("foo": 1, "bar": 2), bar, 3) => ("foo": 1, "bar": 3)
+    #   map-merge((a: (b: c)), a, b, d) => (a: (b: d))
     # @overload map_merge($map1, $map2)
     #   @param $map1 [Sass::Script::Value::Map]
     #   @param $map2 [Sass::Script::Value::Map]
     # @return [Sass::Script::Value::Map]
     # @raise [ArgumentError] if either parameter is not a map
-    def map_merge(map1, map2)
+    # @overload map_merge($map1, $keys..., $map2)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value]]
+    #   @param $map2 [Sass::Script::Value::Map]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if either parameter is not a map
+    def map_merge(map1, *keys)
       assert_type map1, :Map, :map1
+      map2 = keys.pop
+      if map1.is_a? Sass::Script::Value::List
+        map1 = map(map1.to_h)
+      end
```

Comment:
Please do make the changes I suggested. It's always possible to convert a recursive algorithm to an iterative one by using an explicit stack, and doing so will eliminate any risk of stack overflows on large data structures. For example, here's an iterative version of `map_set`:

```ruby
def map_set(map, *keys)
  assert_type map, :Map, :map
  root_map = map.to_h.dup

  current_map = root_map
  keys.each do |key|
    value = current_map[key]
    next_map = value ? value.to_h.dup : {}

    current_map[key] = Map.new(next_map)
    current_map = next_map
  end

  Map.new(root_map)
end
```

Also, you didn't address my other requests: not explicitly checking for a list and getting rid of the `Map.recursive_*` functions.